strategy.py

```diff
@@ -11,94 +11,101 @@ def solve(self, first_field=None):
+        self.game = game
         self.solver = SimplexSolver()
+        self.current_adjacent_fields = []
+
+        self.vars = [[Variable("a[{}][{}]".format(j, i)) for i in range(self.game.board_dim)]
+                                                         for j in range(self.game.board_dim)]
+
+
+    def get_random_field(self):
+        # TODO: we should somehow sample list of safe fields
+        while True:
+            i = randint(0, self.game.board_dim - 1)
+            j = randint(0, self.game.board_dim - 1)
+            if self.vars[i][j].value == 0 and self.game.board[i][j].covered:
+                break
+
+        return self.game.board[i][j]
+
 
     def open(self, field):
-        self.board.open_field(field)
-
-        # this case is really interesting, we are opening a field that is safe but solver
-        # thinks that is is mined. We need to edit and enforce as stay constraint
-        if field.variable.value != 0:
-            field.variable.value = 0
-            self.solver.add_stay(field.variable)
-
-        if field.adjacent_mines == 0:
-            boundary_list = []
-            for adjacent_field in self.board.get_adjacent_fields(field.row, field.column):
-                if adjacent_field.covered:
-                    boundary_list += self.open(adjacent_field)
-            return boundary_list
-        else:
-            return [field]
+        boundary_list = []
+        for field in self.game.open_field(field):
+
+            # this case is really interesting, we are opening a field that is safe but solver
+            # thinks that it is mined. We need to edit and enforce as stay constraint
+            if self.vars[field.row][field.column] != 0:
+                self.vars[field.row][field.column].value = 0
+            self.solver.add_stay(self.vars[field.row][field.column])
+
+            # if adjacent_mines is set we are looking at boundary field and we need to add it
+            # to the boundary_list
+            if field.adjacent_mines:
+                boundary_list.append(field)
+
+            if field in self.current_adjacent_fields:
+                self.current_adjacent_fields.remove(field)
+        return boundary_list
+
 
     def make_constraint(self, row_index, col_index):
         """
         Compute equations for field (row_index, col_index)
         """
 
-        adjacent_fields = self.board.get_adjacent_fields(row_index, col_index)
-        adjacent_mines = self.board.get_adjacent_mines(row_index, col_index)
+        adjacent_fields = self.game.get_adjacent_fields(row_index, col_index)
+        adjacent_mines = self.game.get_adjacent_mines(row_index, col_index)
 
         assert adjacent_mines
         for field in adjacent_fields:
-            if field not in self.board.current_adjacent_fields:
-                self.board.current_adjacent_fields.append(field)
+            if field not in self.current_adjacent_fields and self.game.board[field.row][field.column].covered:
```

Comment:
Tiny optimization that won't include opened fields in current_adjacent_fields and we will have less job to do in solve